addons/mrp/models/mrp_bom.py

```diff
@@ -184,6 +184,37 @@ def _bom_find(self, product_tmpl=None, product=None, picking_type=None, company_
             return domain
         return self.search(domain, order='sequence, product_id', limit=1)
 
+    @api.model
+    def _get_product2bom(self, products, bom_type=False):
+        """Optimized variant of _bom_find to work with recordset"""
+        products = products.filtered(lambda product: product.type != 'service')
+        if not products:
+            return {}
+        product_templates = products.mapped('product_tmpl_id')
+        domain = ['|', ('product_id', 'in', products.ids), '&', ('product_id', '=', False), ('product_tmpl_id', 'in', product_templates.ids)]
+        if self.env.context.get('company_id'):
+            domain = domain + ['|', ('company_id', '=', False), ('company_id', '=', self.env.context.get('company_id'))]
+        if bom_type:
+            domain += [('type', '=', bom_type)]
+
+        boms = self.search(domain, order='sequence, product_id')
```

Comment:
Apparently, it should reversed sorting, because in code below we use the last bom among the options, while in original code it's only the first one  (it has limit=1)

Reply:
@Whenrow could you confirm it?
https://github.com/odoo/odoo/compare/44b33671c45717a507ce28522d9163e1ec8f74ed..5e1b93897bca3f60d5b66d0d889a4c7cf0b906e6

Reply:
Not sure about this one. Building you dict will only get the last bom for a product right ?

Reply:
> Building you dict will only get the last bom for a product right ?

Yes

Reply:
I guess that do the trick even if i'm more partisan of something like 
```python
for bom in boms:
    tmpl = bom.product_tmpl_id
    if tmpl in template2bom:
        template2bom[product] = bom

Reply:
Done via setdefault